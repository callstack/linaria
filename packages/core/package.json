{
  "name": "@linaria/core",
  "version": "4.2.6",
  "description": "Blazing fast zero-runtime CSS in JS library",
  "keywords": [
    "css",
    "css-in-js",
    "linaria",
    "react",
    "styled-components"
  ],
  "homepage": "https://github.com/callstack/linaria#readme",
  "bugs": "https://github.com/callstack/linaria/issues",
  "repository": "git@github.com:callstack/linaria.git",
  "license": "MIT",
  "sideEffects": false,
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "types": "./types/index.d.ts",
      "import": "./dist/index.mjs",
      "default": "./dist/index.js"
    },
    "./*": {
      "types": "./types/*.d.ts",
      "import": "./dist/*.mjs",
      "default": "./dist/*.js"
    }
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "types/index.d.ts",
  "typesVersions": {
    "*": {
      "processors/*": [
        "./types/processors/*.d.ts"
      ]
    }
  },
  "files": [
    "dist/",
    "processors/",
    "types/"
  ],
  "linaria": {
    "tags": {
      "css": "./dist/processors/css.js"
    }
  },
  "scripts": {
    "build": "pnpm build:dist && pnpm build:declarations",
    "build:corejs-test": "cross-env NODE_ENV=legacy babel src --out-dir lib --extensions '.js,.jsx,.ts,.tsx' --ignore \"src/processors/**/*\"",
    "build:declarations": "tsc --emitDeclarationOnly --outDir types",
    "build:dist": "tsup --format cjs,esm",
    "test": "jest --config ../../jest.config.js --rootDir .",
    "test:dts": "dtslint --localTs ../../node_modules/typescript/lib __dtslint__",
    "typecheck": "tsc --noEmit --composite false",
    "watch": "pnpm build:dist --watch & pnpm build:declarations --watch"
  },
  "dependencies": {
    "@linaria/logger": "workspace:^",
    "@linaria/tags": "workspace:^",
    "@linaria/utils": "workspace:^"
  },
  "devDependencies": {
    "@babel/traverse": "^7.20.1",
    "@babel/types": "^7.20.2",
    "@types/babel__core": "^7.1.19",
    "@types/babel__traverse": "^7.17.1",
    "@types/node": "^17.0.39"
  },
  "engines": {
    "node": "^12.16.0 || >=13.7.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "tsup": {
    "entry": [
      "src/index.ts",
      "src/processors/css.ts"
    ],
    "splitting": false,
    "sourcemap": true,
    "clean": true
  }
}
