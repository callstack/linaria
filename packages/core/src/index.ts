export { default as css } from './css';
export { default as cx } from './cx';
export type { CSSProperties } from './CSSProperties';
export type { LinariaClassName } from './cx';
