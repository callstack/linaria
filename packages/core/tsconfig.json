{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "paths": {},
    "rootDir": "src/",
    "types": ["node"]
  },
  "references": [{ "path": "../logger" }, { "path": "../utils" }]
}
