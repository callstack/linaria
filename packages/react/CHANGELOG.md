# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [3.0.0-beta.17](https://github.com/callstack/linaria/compare/v3.0.0-beta.16...v3.0.0-beta.17) (2021-12-27)


### Bug Fixes

* **react:** refactored types for styled function (fixes [#872](https://github.com/callstack/linaria/issues/872)) ([#887](https://github.com/callstack/linaria/issues/887)) ([7b8b129](https://github.com/callstack/linaria/commit/7b8b12937f9a0d1730d908e7cebad1684ccb03c3))





# [3.0.0-beta.15](https://github.com/callstack/linaria/compare/v3.0.0-beta.14...v3.0.0-beta.15) (2021-11-29)


### Bug Fixes

* **react:** fixed types for supporting class components (fixes [#730](https://github.com/callstack/linaria/issues/730)) ([#877](https://github.com/callstack/linaria/issues/877)) ([e637ecb](https://github.com/callstack/linaria/commit/e637ecb8946a8119cfbd039bfb65d42206e09c4e))





# [3.0.0-beta.14](https://github.com/callstack/linaria/compare/v3.0.0-beta.13...v3.0.0-beta.14) (2021-11-05)


### Bug Fixes

* **react:** refactor/rest op ([#860](https://github.com/callstack/linaria/issues/860)) ([da94704](https://github.com/callstack/linaria/commit/da94704df8ca74d94fe57682e2557274cf2d4cb0))
* **react:** unions in prop types are not resolved ([#844](https://github.com/callstack/linaria/issues/844)) ([62009e9](https://github.com/callstack/linaria/commit/62009e9184638fd8761f187c99e7ea434f364bee))





# [3.0.0-beta.13](https://github.com/callstack/linaria/compare/v3.0.0-beta.12...v3.0.0-beta.13) (2021-09-13)


### Bug Fixes

* **react:** fixes for `--exactOptionalPropertyTypes` TS flag ([#827](https://github.com/callstack/linaria/issues/827)) ([eed92b1](https://github.com/callstack/linaria/commit/eed92b19e3b779b656fb780307bbab8a08d14ba2))





# [3.0.0-beta.11](https://github.com/callstack/linaria/compare/v3.0.0-beta.10...v3.0.0-beta.11) (2021-08-08)


### Bug Fixes

* **styled:** remove unnecessary core-js polyfills  (fixes [#799](https://github.com/callstack/linaria/issues/799)) ([#814](https://github.com/callstack/linaria/issues/814)) ([6c3070a](https://github.com/callstack/linaria/commit/6c3070a47715022eb761567b8795f6918784ae4c))





# [3.0.0-beta.7](https://github.com/callstack/linaria/compare/v3.0.0-beta.6...v3.0.0-beta.7) (2021-06-24)

**Note:** Version bump only for package @linaria/react





# [3.0.0-beta.4](https://github.com/callstack/linaria/compare/v3.0.0-beta.3...v3.0.0-beta.4) (2021-05-07)

**Note:** Version bump only for package @linaria/react





# [3.0.0-beta.3](https://github.com/callstack/linaria/compare/v3.0.0-beta.2...v3.0.0-beta.3) (2021-04-20)


### Bug Fixes

* **core,react:** make IE 11 compatible (fixes [#746](https://github.com/callstack/linaria/issues/746)) ([#750](https://github.com/callstack/linaria/issues/750)) ([922df95](https://github.com/callstack/linaria/commit/922df9576a430cdfe9b27aed5dc45c4f75917607))





# [3.0.0-beta.2](https://github.com/callstack/linaria/compare/v3.0.0-beta.1...v3.0.0-beta.2) (2021-04-11)

**Note:** Version bump only for package @linaria/react
