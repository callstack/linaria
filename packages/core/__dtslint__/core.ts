/* eslint-disable @typescript-eslint/no-unused-vars */
import { css, cx } from '../src';

// $ExpectType LinariaClassName
const class1 = css``;

const activeClass = css``;

// $ExpectType string
const combined1 = cx(class1, 'active');

// $ExpectType LinariaClassName
const combined2 = cx(class1, activeClass);
