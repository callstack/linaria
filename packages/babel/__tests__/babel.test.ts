import { join } from 'path';
import { transformAsync } from '@babel/core';
import dedent from 'dedent';
import stripAnsi from 'strip-ansi';
import type { StrictOptions } from '../src';
import serializer from '../__utils__/linaria-snapshot-serializer';

expect.addSnapshotSerializer(serializer);

const transpile = async (
  input: string,
  opts: Partial<StrictOptions> = {
    evaluate: false,
    atomize: require('@linaria/atomic').atomize,
  }
) =>
  (await transformAsync(input, {
    babelrc: false,
    presets: [[require.resolve('../src'), opts]],
    plugins: ['@babel/plugin-syntax-jsx'],
    filename: join(__dirname, 'app/index.js'),
    configFile: false,
  }))!;

it('transpiles styled template literal with object', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    export const Title = styled.h1\`
      font-size: 14px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('uses string passed in as classNameSlug', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    export const Title = styled('h1')\`
      font-size: 14px;
    \`;
`,
    {
      classNameSlug: ['hash', 'title', 'file', 'name', 'ext', 'dir']
        .map((s) => `[${s}]`)
        .join('_'),
    }
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('removes fake replacement patterns in string classNameSlug', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    export const Title = styled('h1')\`
      font-size: 14px;
    \`;
`,
    {
      classNameSlug: `[not]_[actual]_[replacements]`,
    }
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('handles fn passed in as classNameSlug', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    export const Title = styled('h1')\`
      font-size: 14px;
    \`;
`,
    {
      classNameSlug: (hash, title, vars) => {
        return [
          hash,
          title,
          vars.hash,
          vars.title,
          vars.file,
          vars.name,
          vars.ext,
          vars.dir,
        ].join('_');
      },
    }
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('transpiles styled template literal with function and tag', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    export const Title = styled('h1')\`
      font-size: 14px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('transpiles renamed styled import', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled as custom } from '@linaria/react';

    export const Title = custom('h1')\`
      font-size: 14px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('transpiles styled template literal with function and component', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';
    const Heading = () => null;

    export const Title = styled(Heading)\`
      font-size: 14px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('outputs valid CSS classname', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    export const Î©Page$Title = styled.h1\`
      font-size: 14px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('evaluates and inlines expressions in scope', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    const color = 'blue';

    export const Title = styled.h1\`
      color: ${'${color}'};
      width: ${'${100 / 3}'}%;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('inlines object styles as CSS string', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    const cover = {
      '--color-primaryText': '#222',
      position: 'absolute',
      top: 0,
      right: 0,
      bottom: 0,
      left: 0,
      opacity: 1,
      minHeight: 420,

      '&.shouldNotBeChanged': {
        borderColor: '#fff',
      },

      '@media (min-width: 200px)': {
        WebkitOpacity: .8,
        MozOpacity: .8,
        msOpacity: .8,
        OOpacity: .8,
        WebkitBorderRadius: 2,
        MozBorderRadius: 2,
        msBorderRadius: 2,
        OBorderRadius: 2,
        WebkitTransition: '400ms',
        MozTransition: '400ms',
        OTransition: '400ms',
        msTransition: '400ms',
      }
    };

    export const Title = styled.h1\`
      ${'${cover}'}
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('inlines array styles as CSS string', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    const styles = [
      { flex: 1 },
      { display: 'block', height: 24 },
    ];

    export const Title = styled.h1\`
      ${'${styles}'}
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('replaces unknown expressions with CSS custom properties', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    export const Title = styled.h1\`
      font-size: ${'${size}'}px;
      color: ${'${props => props.color}'};
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('handles interpolation followed by unit', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    export const Title = styled.h1\`
      font-size: ${'${size}'}em;
      text-shadow: black 1px ${'${shadow}'}px, white -2px -2px;
      margin: ${'${size}'}px;
      width: calc(2 * ${'${props => props.width}'}vw);
      height: ${'${props => { if (true) { return props.height } else { return 200 } }}'}px;
      grid-template-columns: ${'${unit}'}fr 1fr 1fr ${'${unit}'}fr;
      border-radius: ${'${function(props) { return 200 }}'}px
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('uses the same custom property for the same identifier', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    export const Box = styled.div\`
      height: ${'${size}'}px;
      width: ${'${size}'}px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('uses the same custom property for the same expression', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    export const Box = styled.div\`
      height: ${'${props => props.size}'}px;
      width: ${'${props => props.size}'}px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('handles nested blocks', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    export const Button = styled.button\`
      font-family: ${'${regular}'};

      &:hover {
        border-color: blue;
      }

      @media (max-width: 200px) {
        width: 100%;
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('prevents class name collision', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { styled } from '@linaria/react';

    export const Title = styled.h1\`
      font-size: ${'${size}'}px;
      color: ${'${props => props.color}'}
    \`;

    function Something() {
      const Title = styled.h1\`
        font-family: ${'${regular}'};
      \`;

      return <Title />;
    }
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('does not output CSS if none present', async () => {
  const { code, metadata } = await transpile(
    dedent`
      const number = 42;

      const title = String.raw\`This is something\`;
      `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('does not output CSS property when value is a blank string', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { css } from '@linaria/core';

    export const title = css\`
      font-size: ${''};
      margin: 6px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('transpiles css template literal', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { css } from '@linaria/core';

    export const title = css\`
      font-size: 14px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('handles css template literal in object property', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { css } from '@linaria/core';

    const components = {
      title: css\`
        font-size: 14px;
      \`
    };
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('handles css template literal in JSX element', async () => {
  const { code, metadata } = await transpile(
    dedent`
    import { css } from '@linaria/core';

    <Title class={css\` font-size: 14px; \`} />
    `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('throws when contains dynamic expression without evaluate: true in css tag', async () => {
  expect.assertions(1);

  try {
    await transpile(
      dedent`
      import { css } from '@linaria/core';

      const title = css\`
        font-size: ${'${size}'}px;
      \`;
      `
    );
  } catch (e) {
    expect(
      stripAnsi(e.message.replace(__dirname, '<<DIRNAME>>'))
    ).toMatchSnapshot();
  }
});

it('supports both css and styled tags', async () => {
  const { code, metadata } = await transpile(
    dedent`
      import { css } from '@linaria/core';
      import { styled } from '@linaria/react';

      export const Title = styled.h1\`
        font-size: 14px;
      \`;

      export const title = css\`
        color: blue;
      \`;
      `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('does not include styles if not referenced anywhere', async () => {
  const { code, metadata } = await transpile(
    dedent`
      import { css } from '@linaria/core';
      import { styled } from '@linaria/react';

      const Title = styled.h1\`
        font-size: 14px;
      \`;

      const title = css\`
        color: blue;
      \`;
      `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('includes unreferenced styles for :global', async () => {
  const { code, metadata } = await transpile(
    dedent`
      import { css } from '@linaria/core';
      import { styled } from '@linaria/react';

      const a = css\`
        :global() {
          .title {
            font-size: 14px;
          }
        }
      \`;

      const B = styled.div\`
        :global(.title) {
          font-size: 14px;
        }
      \`;
      `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('handles objects with numeric keys', async () => {
  const { code, metadata } = await transpile(
    dedent`
      import { css } from '@linaria/core';

      export const object = {
        stringKey: css\`\`,
        42: css\`\`,
      }
      `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('handles objects with enums as keys', async () => {
  const { code, metadata } = await transpile(
    dedent`
      import { css } from '@linaria/core';
      import { TestEnum } from './ts-data.ts';

      export const object = {
        [TestEnum.FirstValue]: css\`\`,
        [TestEnum.SecondValue]: css\`\`,
      }
      `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});

it('compiles atomic css', async () => {
  const { code, metadata } = await transpile(
    dedent`
    /* @flow */

    import { css } from '@linaria/atomic';
    import { styled } from '@linaria/react';
    
    const x = css\`
      background: red;
      height: 100px;
    \`;
    
    console.log(x);
    
      `
  );

  expect(code).toMatchSnapshot();
  expect(metadata).toMatchSnapshot();
});
