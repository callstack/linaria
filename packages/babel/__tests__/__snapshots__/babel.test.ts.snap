// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not include styles if not referenced anywhere 1`] = `
"import { css } from '@linaria/core';
import { styled } from '@linaria/react';
const Title = /*#__PURE__*/styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"t17gu1mi\\"
});
const title = \\"tccybe9\\";"
`;

exports[`does not include styles if not referenced anywhere 2`] = `Object {}`;

exports[`does not output CSS if none present 1`] = `
"const number = 42;
const title = String.raw\`This is something\`;"
`;

exports[`does not output CSS if none present 2`] = `Object {}`;

exports[`does not output CSS property when value is a blank string 1`] = `
"import { css } from '@linaria/core';
export const title = \\"t17gu1mi\\";"
`;

exports[`does not output CSS property when value is a blank string 2`] = `

CSS:

.t17gu1mi {
  font-size: ;
  margin: 6px;
}

Dependencies: NA

`;

exports[`evaluates and inlines expressions in scope 1`] = `
"import { styled } from '@linaria/react';
const color = 'blue';
export const Title = /*#__PURE__*/styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"t17gu1mi\\"
});"
`;

exports[`evaluates and inlines expressions in scope 2`] = `

CSS:

.t17gu1mi {
  color: blue;
  width: 33.333333333333336%;
}

Dependencies: NA

`;

exports[`handles css template literal in JSX element 1`] = `
"import { css } from '@linaria/core';
<Title class={\\"t17gu1mi\\"} />;"
`;

exports[`handles css template literal in JSX element 2`] = `

CSS:

.t17gu1mi { font-size: 14px; }

Dependencies: NA

`;

exports[`handles css template literal in object property 1`] = `
"import { css } from '@linaria/core';
const components = {
  title: \\"t17gu1mi\\"
};"
`;

exports[`handles css template literal in object property 2`] = `

CSS:

.t17gu1mi {
    font-size: 14px;
  }

Dependencies: NA

`;

exports[`handles fn passed in as classNameSlug 1`] = `
"import { styled } from '@linaria/react';
export const Title = /*#__PURE__*/styled('h1')({
  name: \\"Title\\",
  class: \\"t17gu1mi_Title_t17gu1mi_Title__tests___app_index_js_index__js_app\\"
});"
`;

exports[`handles fn passed in as classNameSlug 2`] = `

CSS:

.t17gu1mi_Title_t17gu1mi_Title__tests___app_index_js_index__js_app {
  font-size: 14px;
}

Dependencies: NA

`;

exports[`handles interpolation followed by unit 1`] = `
"import { styled } from '@linaria/react';
export const Title = /*#__PURE__*/styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"t17gu1mi\\",
  vars: {
    \\"t17gu1mi-0\\": [size, \\"em\\"],
    \\"t17gu1mi-1\\": [shadow, \\"px\\"],
    \\"t17gu1mi-2\\": [size, \\"px\\"],
    \\"t17gu1mi-3\\": [props => props.width, \\"vw\\"],
    \\"t17gu1mi-4\\": [props => {
      if (true) {
        return props.height;
      } else {
        return 200;
      }
    }, \\"px\\"],
    \\"t17gu1mi-5\\": [unit, \\"fr\\"],
    \\"t17gu1mi-7\\": [function (props) {
      return 200;
    }, \\"px\\"]
  }
});"
`;

exports[`handles interpolation followed by unit 2`] = `

CSS:

.t17gu1mi {
  font-size: var(--t17gu1mi-0);
  text-shadow: black 1px var(--t17gu1mi-1), white -2px -2px;
  margin: var(--t17gu1mi-2);
  width: calc(2 * var(--t17gu1mi-3));
  height: var(--t17gu1mi-4);
  grid-template-columns: var(--t17gu1mi-5) 1fr 1fr var(--t17gu1mi-5);
  border-radius: var(--t17gu1mi-7)
}

Dependencies: NA

`;

exports[`handles nested blocks 1`] = `
"import { styled } from '@linaria/react';
export const Button = /*#__PURE__*/styled(\\"button\\")({
  name: \\"Button\\",
  class: \\"b17gu1mi\\",
  vars: {
    \\"b17gu1mi-0\\": [regular]
  }
});"
`;

exports[`handles nested blocks 2`] = `

CSS:

.b17gu1mi {
  font-family: var(--b17gu1mi-0);

  &:hover {
    border-color: blue;
  }

  @media (max-width: 200px) {
    width: 100%;
  }
}

Dependencies: NA

`;

exports[`handles objects with enums as keys 1`] = `
"import { css } from '@linaria/core';
import { TestEnum } from './ts-data.ts';
export const object = {
  [TestEnum.FirstValue]: \\"t17gu1mi\\",
  [TestEnum.SecondValue]: \\"tccybe9\\"
};"
`;

exports[`handles objects with enums as keys 2`] = `

CSS:

.t17gu1mi {}
.tccybe9 {}

Dependencies: NA

`;

exports[`handles objects with numeric keys 1`] = `
"import { css } from '@linaria/core';
export const object = {
  stringKey: \\"s17gu1mi\\",
  42: \\"_ccybe9\\"
};"
`;

exports[`handles objects with numeric keys 2`] = `

CSS:

.s17gu1mi {}
._ccybe9 {}

Dependencies: NA

`;

exports[`includes unreferenced styles for :global 1`] = `
"import { css } from '@linaria/core';
import { styled } from '@linaria/react';
const a = \\"a17gu1mi\\";
const B = /*#__PURE__*/styled(\\"div\\")({
  name: \\"B\\",
  class: \\"bccybe9\\"
});"
`;

exports[`includes unreferenced styles for :global 2`] = `

CSS:

.a17gu1mi {
  :global() {
    .title {
      font-size: 14px;
    }
  }
}
.bccybe9 {
  :global(.title) {
    font-size: 14px;
  }
}

Dependencies: NA

`;

exports[`inlines array styles as CSS string 1`] = `
"import { styled } from '@linaria/react';
const styles = [{
  flex: 1
}, {
  display: 'block',
  height: 24
}];
export const Title = /*#__PURE__*/styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"t17gu1mi\\"
});"
`;

exports[`inlines array styles as CSS string 2`] = `

CSS:

.t17gu1mi {
  flex: 1; display: block; height: 24px;
}

Dependencies: NA

`;

exports[`inlines object styles as CSS string 1`] = `
"import { styled } from '@linaria/react';
const cover = {
  '--color-primaryText': '#222',
  position: 'absolute',
  top: 0,
  right: 0,
  bottom: 0,
  left: 0,
  opacity: 1,
  minHeight: 420,
  '&.shouldNotBeChanged': {
    borderColor: '#fff'
  },
  '@media (min-width: 200px)': {
    WebkitOpacity: .8,
    MozOpacity: .8,
    msOpacity: .8,
    OOpacity: .8,
    WebkitBorderRadius: 2,
    MozBorderRadius: 2,
    msBorderRadius: 2,
    OBorderRadius: 2,
    WebkitTransition: '400ms',
    MozTransition: '400ms',
    OTransition: '400ms',
    msTransition: '400ms'
  }
};
export const Title = /*#__PURE__*/styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"t17gu1mi\\"
});"
`;

exports[`inlines object styles as CSS string 2`] = `

CSS:

.t17gu1mi {
  --color-primaryText: #222; position: absolute; top: 0; right: 0; bottom: 0; left: 0; opacity: 1; min-height: 420px; &.shouldNotBeChanged { border-color: #fff; } @media (min-width: 200px) { -webkit-opacity: 0.8; -moz-opacity: 0.8; -ms-opacity: 0.8; -o-opacity: 0.8; -webkit-border-radius: 2px; -moz-border-radius: 2px; -ms-border-radius: 2px; -o-border-radius: 2px; -webkit-transition: 400ms; -moz-transition: 400ms; -o-transition: 400ms; -ms-transition: 400ms; }
}

Dependencies: NA

`;

exports[`outputs valid CSS classname 1`] = `
"import { styled } from '@linaria/react';
export const ΩPage$Title = /*#__PURE__*/styled(\\"h1\\")({
  name: \\"\\\\u03A9Page$Title\\",
  class: \\"\\\\u03C917gu1mi\\"
});"
`;

exports[`outputs valid CSS classname 2`] = `

CSS:

.ω17gu1mi {
  font-size: 14px;
}

Dependencies: NA

`;

exports[`prevents class name collision 1`] = `
"import { styled } from '@linaria/react';
export const Title = /*#__PURE__*/styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"t17gu1mi\\",
  vars: {
    \\"t17gu1mi-0\\": [size, \\"px\\"],
    \\"t17gu1mi-1\\": [props => props.color]
  }
});

function Something() {
  const Title = /*#__PURE__*/styled(\\"h1\\")({
    name: \\"Title\\",
    class: \\"tccybe9\\",
    vars: {
      \\"tccybe9-0\\": [regular]
    }
  });
  return <Title />;
}"
`;

exports[`prevents class name collision 2`] = `

CSS:

.t17gu1mi {
  font-size: var(--t17gu1mi-0);
  color: var(--t17gu1mi-1)
}
.tccybe9 {
    font-family: var(--tccybe9-0);
  }

Dependencies: NA

`;

exports[`removes fake replacement patterns in string classNameSlug 1`] = `
"import { styled } from '@linaria/react';
export const Title = /*#__PURE__*/styled('h1')({
  name: \\"Title\\",
  class: \\"__\\"
});"
`;

exports[`removes fake replacement patterns in string classNameSlug 2`] = `

CSS:

.__ {
  font-size: 14px;
}

Dependencies: NA

`;

exports[`replaces unknown expressions with CSS custom properties 1`] = `
"import { styled } from '@linaria/react';
export const Title = /*#__PURE__*/styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"t17gu1mi\\",
  vars: {
    \\"t17gu1mi-0\\": [size, \\"px\\"],
    \\"t17gu1mi-1\\": [props => props.color]
  }
});"
`;

exports[`replaces unknown expressions with CSS custom properties 2`] = `

CSS:

.t17gu1mi {
  font-size: var(--t17gu1mi-0);
  color: var(--t17gu1mi-1);
}

Dependencies: NA

`;

exports[`supports both css and styled tags 1`] = `
"import { css } from '@linaria/core';
import { styled } from '@linaria/react';
export const Title = /*#__PURE__*/styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"t17gu1mi\\"
});
export const title = \\"tccybe9\\";"
`;

exports[`supports both css and styled tags 2`] = `

CSS:

.t17gu1mi {
  font-size: 14px;
}
.tccybe9 {
  color: blue;
}

Dependencies: NA

`;

exports[`throws when contains dynamic expression without evaluate: true in css tag 1`] = `
"<<DIRNAME>>/app/index.js: The CSS cannot contain JavaScript expressions when using the 'css' tag. To evaluate the expressions at build time, pass 'evaluate: true' to the babel plugin.
  2 |
  3 | const title = css\`
> 4 |   font-size: \${size}px;
    |                ^^^^
  5 | \`;"
`;

exports[`transpiles css template literal 1`] = `
"import { css } from '@linaria/core';
export const title = \\"t17gu1mi\\";"
`;

exports[`transpiles css template literal 2`] = `

CSS:

.t17gu1mi {
  font-size: 14px;
}

Dependencies: NA

`;

exports[`transpiles renamed styled import 1`] = `
"import { styled as custom } from '@linaria/react';
export const Title = /*#__PURE__*/custom('h1')({
  name: \\"Title\\",
  class: \\"t17gu1mi\\"
});"
`;

exports[`transpiles renamed styled import 2`] = `

CSS:

.t17gu1mi {
  font-size: 14px;
}

Dependencies: NA

`;

exports[`transpiles styled template literal with function and component 1`] = `
"import { styled } from '@linaria/react';

const Heading = () => null;

export const Title = /*#__PURE__*/styled(Heading)({
  name: \\"Title\\",
  class: \\"t17gu1mi\\"
});"
`;

exports[`transpiles styled template literal with function and component 2`] = `

CSS:

.t17gu1mi {
  font-size: 14px;
}

Dependencies: NA

`;

exports[`transpiles styled template literal with function and tag 1`] = `
"import { styled } from '@linaria/react';
export const Title = /*#__PURE__*/styled('h1')({
  name: \\"Title\\",
  class: \\"t17gu1mi\\"
});"
`;

exports[`transpiles styled template literal with function and tag 2`] = `

CSS:

.t17gu1mi {
  font-size: 14px;
}

Dependencies: NA

`;

exports[`transpiles styled template literal with object 1`] = `
"import { styled } from '@linaria/react';
export const Title = /*#__PURE__*/styled(\\"h1\\")({
  name: \\"Title\\",
  class: \\"t17gu1mi\\"
});"
`;

exports[`transpiles styled template literal with object 2`] = `

CSS:

.t17gu1mi {
  font-size: 14px;
}

Dependencies: NA

`;

exports[`uses string passed in as classNameSlug 1`] = `
"import { styled } from '@linaria/react';
export const Title = /*#__PURE__*/styled('h1')({
  name: \\"Title\\",
  class: \\"t17gu1mi_Title__tests___app_index_js_index__js_app\\"
});"
`;

exports[`uses string passed in as classNameSlug 2`] = `

CSS:

.t17gu1mi_Title__tests___app_index_js_index__js_app {
  font-size: 14px;
}

Dependencies: NA

`;

exports[`uses the same custom property for the same expression 1`] = `
"import { styled } from '@linaria/react';
export const Box = /*#__PURE__*/styled(\\"div\\")({
  name: \\"Box\\",
  class: \\"b17gu1mi\\",
  vars: {
    \\"b17gu1mi-0\\": [props => props.size, \\"px\\"]
  }
});"
`;

exports[`uses the same custom property for the same expression 2`] = `

CSS:

.b17gu1mi {
  height: var(--b17gu1mi-0);
  width: var(--b17gu1mi-0);
}

Dependencies: NA

`;

exports[`uses the same custom property for the same identifier 1`] = `
"import { styled } from '@linaria/react';
export const Box = /*#__PURE__*/styled(\\"div\\")({
  name: \\"Box\\",
  class: \\"b17gu1mi\\",
  vars: {
    \\"b17gu1mi-0\\": [size, \\"px\\"]
  }
});"
`;

exports[`uses the same custom property for the same identifier 2`] = `

CSS:

.b17gu1mi {
  height: var(--b17gu1mi-0);
  width: var(--b17gu1mi-0);
}

Dependencies: NA

`;
